sympy/core/exprtools.py

```diff
@@ -1414,156 +1288,614 @@ def factor_nc(expr):
         return expr.func(*[factor_nc(a) for a in expr.args])
     expr = expr.func(*[expand_power_exp(i) for i in expr.args])
 
-    from sympy.polys.polytools import gcd, factor
-    expr, rep, nc_symbols = _mask_nc(expr)
+    return _NCPoly(expr).factor_one()
 
-    if rep:
-        return factor(expr).subs(rep)
-    else:
-        args = [a.args_cnc() for a in Add.make_args(expr)]
-        c = g = l = r = S.One
-        hit = False
-        # find any commutative gcd term
-        for i, a in enumerate(args):
-            if i == 0:
-                c = Mul._from_args(a[0])
-            elif a[0]:
-                c = gcd(c, Mul._from_args(a[0]))
+
+def _nc_generators(expr):
+    """Return noncommutative polynomial generators of an expression.
+
+    Commutative symbols are excluded, and functions of generators other than
+    positive integer powers are treated as generators, as with the polynomials module
+    """
+    if expr.is_Add or expr.is_Mul:
+        generators = set()
+        for arg in expr.args:
+            generators |= set(_nc_generators(arg))
+        return generators
+    if expr.is_commutative:
+        return set()
+    if expr.is_Pow:
+        term, pw = expr.args
+        if pw.is_positive and pw.is_Integer:
+            return _nc_generators(term)
+    return {expr}
+
+def _nc_extensions(expr):
+    """Return commutative non-rational field extensions of a non-commutative polynomial's coefficients
+    """
+    if expr.is_rational:
+        return set()
+    if expr.is_Add or expr.is_Mul:
+        extensions = set()
+        for arg in expr.args:
+            extensions |= _nc_extensions(arg)
+        return extensions
+    if expr.is_Pow:
+        term, pw = expr.args
+        if pw.is_Integer:
+            return _nc_extensions(term)
+    return {expr}
+
+def _valid_nc_coeff(coeff, extensions):
+    """Return whether the given coefficient is valid given the extensions"""
+    if coeff.is_rational:
+        return True
+    if coeff in extensions:
+        return True
+    if coeff.is_Add or coeff.is_Mul:
+        for arg in coeff.args:
+            if not _valid_nc_coeff(arg, extensions):
+                return False
+        return True
+    if coeff.is_Pow:
+        term, pw = coeff.args
+        if pw.is_Integer:
+            return _valid_nc_coeff(term, extensions)
+    return False
+
+
+class _NCMonomial:
+    """Class that represents a monomial with noncommutative generators"""
+    def __init__(self, coeff, term, generators, degree=None):
+        self.coeff = coeff
+        self.term = term
```

Comment:
The terminology here is confusing. If we have a coefficient then I would say that this class represents a "term" rather than a "monomial".

Maybe it would make more sense for the coefficient to be separate so that this is a monomial or maybe just the class should be renamed.

Reply:
I don't object to renaming the class or considering keeping the coefficient separate, but a monomial doesn't necessarily exclude the coefficient: https://en.wikipedia.org/wiki/Monomial

Reply:
The terminology as used in sympy is quite consistent about the difference between coefficients, monomials and terms so we should preserve that.

Reply:
Ah, looking at the sympy poly codebase I see that `monomial` and `term` have well-established meanings. Makes sense